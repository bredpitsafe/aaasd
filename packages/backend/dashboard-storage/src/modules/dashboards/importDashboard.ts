import type { UnaryRpc } from '@backend/grpc/src/types/rpc-types.ts';
import type { TPermission } from '@grpc-schemas/dashboard_storage-api-sdk/services/dashboard_storage/v1/common';
import type {
    TImportDashboardRequest,
    TImportDashboardResponse,
} from '@grpc-schemas/dashboard_storage-api-sdk/services/dashboard_storage/v1/dashboard_api';
import { randomUUID } from 'crypto';
import { intersection, isNil } from 'lodash-es';
import { map, of, switchMap } from 'rxjs';
import { tap } from 'rxjs/operators';

import { getDashboardsByDigest, upsertDashboard } from '../../db/dashboards.ts';
import type { TUpdatableGrpcPermissionRow } from '../../db/permissions.ts';
import { getPermissions, upsertPermissions } from '../../db/permissions.ts';
import { getDigest } from '../../db/utils.ts';
import { EActorName } from '../../def/actor.ts';
import { handleUnaryRpcInitialization } from '../handlers/handleUnaryRpcInitialization.ts';

export const importDashboard: UnaryRpc<TImportDashboardRequest, TImportDashboardResponse> =
    function importDashboard(call, callback) {
        const { logger, messageWithContext, username, logStart, payload, handle$ } =
            handleUnaryRpcInitialization({
                rpc: importDashboard,
                call,
                callback,
                actor: EActorName.Dashboards,
            });

        logStart();

        const grpcPermissionRow: TUpdatableGrpcPermissionRow = {
            user: username,
            permission: 'PERMISSION_OWNER',
        };

        handle$({
            obs$: of(undefined).pipe(
                switchMap(async () => {
                    // Deduplicate existing dashboards by digest (per user)
                    const digest = getDigest(payload.config);
                    const dashboardIds = (await getDashboardsByDigest(digest)).map((d) => d.id);
                    const allowedLevels: TPermission[] = [
                        'PERMISSION_OWNER',
                        'PERMISSION_EDITOR',
                        'PERMISSION_VIEWER',
                    ];
                    const permissionDashboardIds = (
                        await getPermissions({ allowedLevels, user: username })
                    ).map((p) => p.dashboardId);
                    let id: string | undefined = intersection(
                        dashboardIds,
                        permissionDashboardIds,
                    )[0];

                    if (!isNil(id)) {
                        logger.info({
                            message: messageWithContext('dashboard duplicate found by digest'),
                            id,
                            digest,
                        });

                        return Promise.resolve(id);
                    }

                    id = randomUUID();

                    logger.info({
                        message: messageWithContext('creating a new dashboard'),
                        id,
                    });

                    return upsertDashboard({ ...payload, status: 'STATUS_ACTIVE', id });
                }),
                tap((id) => {
                    logger.info({
                        message: messageWithContext('dashboard created if not exists'),
                        id,
                    });
                }),
                switchMap((id) => upsertPermissions(id, [grpcPermissionRow])),
                map((id) => {
                    logger.info({
                        message: messageWithContext('permissions upserted'),
                        id,
                    });

                    return { id };
                }),
            ),
        });
    };
