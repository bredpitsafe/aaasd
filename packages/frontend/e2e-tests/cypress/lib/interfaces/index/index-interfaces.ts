export enum EPagesName {
    dashboard = 'Dashboard',
    tradingServersManager = 'Trading Servers Manager',
    herodotusTerminal = 'Herodotus Terminal',
    herodotusTrades = 'Herodotus Trades',
    herodotusDashboard = 'Herodotus Dashboard',
    tradingStats = 'Trading Stats',
    backtestingManager = 'Backtesting Manager',
    jsonCharts = 'JSON Charts',
    balanceMonitor = 'Balance Monitor',
    wsQueryTerminal = 'WS Query Terminal',
}
