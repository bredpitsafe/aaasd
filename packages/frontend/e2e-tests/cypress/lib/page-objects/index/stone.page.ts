import { BasePage } from '../../base.page';
import { Button } from '../../base/elements/button';
import { Text } from '../../base/elements/text';
enum EStonePageSelectors {
    Dashboard = '[href*=dashboard]',
    TradingServersManager = '[href*=trading-servers-manager]',
    HerodotusTerminal = '[href*=herodotus-terminal]',
    TradingStats = '[href*=trading-stats]',
    BacktestingManager = '[href*=backtesting]',
    BalanceMonitor = '[href*=balance-monitor]',
    WSQueryTerminal = '[href*=ws-query-terminal]',
    JsonCharts = '[href*=charter-json-viewer]',
    LoadingIcon = '[class*="anticon anticon-loading"]',
}

class StonePage extends BasePage {
    readonly dashboard = new Text(EStonePageSelectors.Dashboard, false);
    readonly tradingServersManager = new Text(EStonePageSelectors.TradingServersManager, false);
    readonly herodotusTerminal = new Text(EStonePageSelectors.HerodotusTerminal, false);
    readonly tradingStats = new Text(EStonePageSelectors.TradingStats, false);
    readonly backtestingManager = new Text(EStonePageSelectors.BacktestingManager, false);
    readonly balanceMonitor = new Text(EStonePageSelectors.BalanceMonitor, false);
    readonly wsQueryTerminal = new Text(EStonePageSelectors.WSQueryTerminal, false);
    readonly jsonCharts = new Text(EStonePageSelectors.JsonCharts, false);
    readonly loadingIcon = new Button(EStonePageSelectors.LoadingIcon, false);

    checkElementsExists(): void {
        this.dashboard.checkExists();
        this.tradingServersManager.checkExists();
        this.herodotusTerminal.checkExists();
        this.tradingStats.checkExists();
        this.backtestingManager.checkExists();
        this.backtestingManager.checkExists();
    }

    open(): StonePage {
        this.visit();
        return this;
    }
}

export const stonePage = new StonePage();
