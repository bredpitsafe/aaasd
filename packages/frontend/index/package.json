{
  "name": "@frontend/index",
  "version": "1.0.0",
  "description": "simple index page for choose application",
  "scripts": {
    "build": "NODE_ENV=production npx webpack",
    "lint": "eslint './src/**/*.{js,ts,tsx}' --quiet",
    "server": "npx webpack serve",
    "test_unit": "jest --passWithNoTests",
    "types": "tsc --noEmit"
  },
  "dependencies": {
    "@frontend/common": "^1.0.1",
    "@common/types": "^1.0.0"
  }
}
