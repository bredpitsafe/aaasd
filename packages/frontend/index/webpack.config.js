const {
    buildConfig,
    htmlConfig,
    defaultInOut,
    defaultDevServerConfig,
} = require('../../../webpack/configs');
const path = require('path');

const app = 'index';
const resolvePath = () => path.resolve(__dirname);

module.exports = buildConfig([
    defaultInOut([app], resolvePath),
    htmlConfig([app], resolvePath),
    defaultDevServerConfig(app),
]);
