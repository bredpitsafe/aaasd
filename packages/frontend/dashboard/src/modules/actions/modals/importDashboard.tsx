import { assertNever } from '@common/utils/src/assert.ts';
import type { TBase64 } from '@common/utils/src/base64.ts';
import { base64ToString } from '@common/utils/src/base64.ts';
import { ModuleModals } from '@frontend/common/src/lib/modals';
import type {
    TStorageDashboardConfig,
    TStorageDashboardName,
} from '@frontend/common/src/types/domain/dashboardsStorage';
import { EGrpcErrorCode, GrpcError } from '@frontend/common/src/types/GrpcError.ts';
import { EDataType } from '@frontend/common/src/utils/dataFormat';
import { Defer } from '@frontend/common/src/utils/Defer.ts';
import { createObservableProcedure } from '@frontend/common/src/utils/LPC/createObservableProcedure.ts';
import {
    mapValueDescriptor,
    squashValueDescriptors,
    switchMapValueDescriptor,
    takeWhileFirstSyncValueDescriptor,
} from '@frontend/common/src/utils/Rx/ValueDescriptor2.ts';
import { createSyncedValueDescriptor } from '@frontend/common/src/utils/ValueDescriptor/utils';
import { isNil } from 'lodash-es';
import { combineLatest, from, Observable, of, switchMap } from 'rxjs';
import { first, map } from 'rxjs/operators';

import type { TDashboard } from '../../../types/dashboard';
import type { TImportableDashboard } from '../../../types/dashboard/importable';
import type { TDashboardItemKey } from '../../../types/fullDashboard';
import { EOpenType } from '../../../types/router';
import { getDashboardItemKeyFromDashboard, unifyIds } from '../../../utils/dashboards';
import {
    convertDashboardToExportableDashboardEditor,
    convertDashboardToXml,
    convertImportableXmlDashboardToDashboard,
    convertXMLToDashboard,
    validateFullDashboardJsonStructure,
} from '../../../utils/dashboards/converters';
import { ModuleUI } from '../../ui/module';
import { ModuleCreateDashboard } from '../dashboards/createDashboard';
import { ModuleUpdateDashboard } from '../dashboards/updateDashboard';
import { ModuleGetDashboardList } from '../fullDashboards/getDashboardList';
import { ModuleSubscribeToDashboard } from '../fullDashboards/ModuleSubscribeToDashboard.ts';
import { ModuleNavigateByDashboardItemKey } from '../navigation/navigateByDashboardItemKey.ts';

export const ModuleImportDashboard = createObservableProcedure((ctx) => {
    const importFullDashboardFromBase64 = ModuleImportFullDashboardFromBase64(ctx);
    const importFullDashboardFromXmlString = ModuleImportFullDashboardFromXmlString(ctx);
    const importFullDashboardFromJsonString = ModuleImportFullDashboardFromJsonString(ctx);

    return (
        {
            type,
            config,
            confirmAdd,
        }: {
            type: EDataType;
            config: string;
            confirmAdd?: boolean;
        },
        options,
    ) => {
        if (type === EDataType.XML) {
            return importFullDashboardFromXmlString(
                {
                    config: config as TStorageDashboardConfig,
                    confirmAdd,
                },
                options,
            );
        }

        if (type === EDataType.JSON) {
            return importFullDashboardFromJsonString({ config, confirmAdd }, options);
        }

        if (type === EDataType.Base64) {
            return importFullDashboardFromBase64(
                {
                    config: config as TBase64<TStorageDashboardConfig>,
                    confirmAdd,
                },
                options,
            );
        }

        assertNever(type);
    };
});

const ModuleImportFullDashboardFromBase64 = createObservableProcedure((ctx) => {
    const importFullDashboardFromJsonString = ModuleImportFullDashboardFromJsonString(ctx);

    return (
        {
            config,
            confirmAdd,
        }: {
            config: TBase64<string>;
            confirmAdd?: boolean;
        },
        options,
    ) => {
        try {
            return importFullDashboardFromJsonString(
                { config: base64ToString(config), confirmAdd },
                options,
            );
        } catch (e) {
            throw new Error('Could not parse dashboard base64');
        }
    };
});

const ModuleImportFullDashboardFromJsonString = createObservableProcedure((ctx) => {
    const importFullDashboardFromJson = ModuleImportFullDashboardFromJson(ctx);

    return ({ config, confirmAdd }: { config: string; confirmAdd?: boolean }, options) => {
        return new Observable<TImportableDashboard>((subscriber) => {
            subscriber.next(JSON.parse(config));
            subscriber.complete();
        }).pipe(
            switchMap((json) => {
                return from(validateFullDashboardJsonStructure(json)).pipe(
                    map((result) => {
                        if (result) return json;
                        throw new GrpcError('Incorrect Dashboard JSON', {
                            code: EGrpcErrorCode.INVALID_ARGUMENT,
                        });
                    }),
                );
            }),
            switchMap((json) => {
                const dashboard = convertImportableXmlDashboardToDashboard(json);
                return importFullDashboardFromJson(
                    { dashboard: unifyIds(dashboard), confirmAdd },
                    options,
                );
            }),
        );
    };
});

const ModuleImportFullDashboardFromXmlString = createObservableProcedure((ctx) => {
    const importFullDashboardFromJson = ModuleImportFullDashboardFromJson(ctx);

    return (
        { config, confirmAdd }: { config: TStorageDashboardConfig; confirmAdd?: boolean },
        options,
    ) => {
        return from(convertXMLToDashboard(config)).pipe(
            switchMap((dashboard) => {
                return importFullDashboardFromJson(
                    { dashboard: unifyIds(dashboard), confirmAdd },
                    options,
                );
            }),
        );
    };
});

const ModuleImportFullDashboardFromJson = createObservableProcedure((ctx) => {
    const { show } = ModuleModals(ctx);
    const { currentDashboardItemKey$ } = ModuleUI(ctx);
    const getDashboardList = ModuleGetDashboardList(ctx);
    const resolveDashboard = ModuleResolveDashboard(ctx);

    return (
        {
            dashboard,
            confirmAdd,
        }: {
            dashboard: TDashboard;
            confirmAdd?: boolean;
        },
        options,
    ) => {
        if (confirmAdd !== true) {
            return resolveDashboard(
                {
                    name: dashboard.name,
                    dashboard,
                },
                options,
            );
        }

        return combineLatest([
            currentDashboardItemKey$.pipe(first(), map(createSyncedValueDescriptor)),
            getDashboardList(undefined, options),
            import('../../../components/modals/SelectDashboardNameDialog').then((m) =>
                createSyncedValueDescriptor(m.SelectDashboardNameDialog),
            ),
        ]).pipe(
            squashValueDescriptors(),
            takeWhileFirstSyncValueDescriptor(),
            switchMapValueDescriptor(({ value }) => {
                const defer = new Defer<
                    | undefined
                    | {
                          id?: TDashboardItemKey;
                          name: TStorageDashboardName;
                          dashboard: TDashboard;
                          openType?: EOpenType;
                      }
                >();
                const [currentDashboardItemKey, dashboardItems, SelectDashboardNameDialog] = value;
                const handleSet = (
                    id: TDashboardItemKey | undefined,
                    name: TStorageDashboardName,
                    openType: EOpenType,
                ) => {
                    modal.destroy();
                    defer.resolve({
                        id,
                        name,
                        openType,
                        dashboard,
                    });
                };
                const handleCancel = () => {
                    modal.destroy();
                    defer.resolve(undefined);
                };

                const modal = show(
                    <SelectDashboardNameDialog
                        title="Add dashboard"
                        name={dashboard.name}
                        dashboardItems={dashboardItems}
                        allowToChooseNewTab={!isNil(currentDashboardItemKey)}
                        onSet={handleSet}
                        onCancel={handleCancel}
                    />,
                );

                return from(defer.promise.then(createSyncedValueDescriptor));
            }),
            switchMapValueDescriptor(({ value }) => {
                return isNil(value)
                    ? of(createSyncedValueDescriptor(false))
                    : resolveDashboard(value, options);
            }),
        );
    };
});

// TODO: give better name/split to 2 functions
const ModuleResolveDashboard = createObservableProcedure((ctx) => {
    const subscribeToDashboard = ModuleSubscribeToDashboard(ctx);
    const updateDashboard = ModuleUpdateDashboard(ctx);
    const createDashboard = ModuleCreateDashboard(ctx);
    const navigateByDashboardItemKey = ModuleNavigateByDashboardItemKey(ctx);

    return (
        {
            id,
            name,
            dashboard,
            openType = EOpenType.CurrentWindow,
        }: {
            id?: TDashboardItemKey;
            name: TStorageDashboardName;
            dashboard: TDashboard;
            openType?: EOpenType;
        },
        options,
    ) => {
        if (isNil(id)) {
            return createDashboard(
                {
                    name,
                    config: convertDashboardToXml(
                        convertDashboardToExportableDashboardEditor(dashboard),
                    ),
                },
                options,
            ).pipe(
                switchMapValueDescriptor(({ value: key }) => {
                    return navigateByDashboardItemKey({ key, openType }, options).pipe(
                        map(createSyncedValueDescriptor),
                    );
                }),
                mapValueDescriptor(() => createSyncedValueDescriptor(true)),
            );
        } else {
            return subscribeToDashboard(id, options).pipe(
                takeWhileFirstSyncValueDescriptor(),
                switchMapValueDescriptor(({ value }) => {
                    return updateDashboard(
                        {
                            ...value,
                            dashboard: { ...dashboard, name },
                        },
                        options,
                    ).pipe(mapValueDescriptor(() => createSyncedValueDescriptor(value)));
                }),
                switchMapValueDescriptor(({ value }) => {
                    return navigateByDashboardItemKey(
                        {
                            key: getDashboardItemKeyFromDashboard(value),
                            openType,
                        },
                        options,
                    ).pipe(map(() => createSyncedValueDescriptor(true)));
                }),
            );
        }
    };
});
