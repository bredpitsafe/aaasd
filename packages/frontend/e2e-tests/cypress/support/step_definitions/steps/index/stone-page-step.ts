import { Given } from '@badeball/cypress-cucumber-preprocessor';

import { EPagesName } from '../../../../lib/interfaces/index/index-interfaces';
import { ERobots } from '../../../../lib/interfaces/trading-servers-manager/robots-interfaces';
import { backtestingPage } from '../../../../lib/page-objects/backtesting/backtesting.page';
import { balanceMonitorPage } from '../../../../lib/page-objects/balance-monitor/balance-monitor.page';
import { charterJsonViewerPage } from '../../../../lib/page-objects/charter-json-viewer/charter-json-viewer.page';
import { authenticationPage } from '../../../../lib/page-objects/common/authentication.page';
import { settingsModal } from '../../../../lib/page-objects/common/settings.modal';
import { tableRow } from '../../../../lib/page-objects/common/table/table.row';
import { dashboardPage } from '../../../../lib/page-objects/dashboard/dashboard.page';
import { herodotusTerminalPage } from '../../../../lib/page-objects/herodotus-terminal/herodotus-terminal.page';
import { herodotusTradesPage } from '../../../../lib/page-objects/herodotus-trades/herodotus-trades.page';
import { stonePage } from '../../../../lib/page-objects/index/stone.page';
import { tradingServersManagerPage } from '../../../../lib/page-objects/trading-servers-manager/trading-servers-manager.page';
import { tradingStatsPage } from '../../../../lib/page-objects/trading-stats/trading-stats.page';
import { wsQueryTerminalPage } from '../../../../lib/page-objects/ws-query-terminal/ws-query-terminal.page';
import { customWait } from '../../../../lib/web-socket/server';
import { checkUrlInclude } from '../../../asserts/comon-url-assert';

Given(`user goes to the "Stone" page`, () => {
    stonePage.visit();
    stonePage.checkElementsExists();
});

Given(`user click an the {string} link`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.dashboard:
            stonePage.dashboard.click();
            break;
        case EPagesName.tradingServersManager:
            stonePage.tradingServersManager.click();
            break;
        case EPagesName.herodotusTerminal:
            stonePage.herodotusTerminal.click();
            break;
        case EPagesName.tradingStats:
            stonePage.tradingStats.click();
            break;
        case EPagesName.backtestingManager:
            stonePage.backtestingManager.click();
            break;
        case EPagesName.jsonCharts:
            stonePage.jsonCharts.click();
            break;
        case EPagesName.balanceMonitor:
            stonePage.balanceMonitor.click();
            break;
        case EPagesName.wsQueryTerminal:
            stonePage.wsQueryTerminal.click();
            break;
    }
});

Given(
    `user on the {string} page with the "default" backend server parameter`,
    (pageName: EPagesName) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.checkElementsExists();
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.checkElementsExists();
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.checkElementsExists();
                break;
            case EPagesName.backtestingManager:
                backtestingPage.checkElementsExists();
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.checkElementsExists();
                break;
        }
    },
);

Given(`user not sees data on the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.checkNotVisibleTabPanel();
            tradingServersManagerPage.checkVisibleLoading();
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.checkVisibleTabPanel();
            herodotusTerminalPage.checkTaskTableVisibility(false);
            stonePage.loadingIcon.checkVisible();
            break;
        case EPagesName.backtestingManager:
            backtestingPage.checkNotExistsTask();
            break;
    }
});

Given(`user sees the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.checkVisibleTabPanel();
            checkUrlInclude(tradingServersManagerPage.pageUrl);
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.checkVisibleTabPanel();
            herodotusTerminalPage.checkTaskTableVisibility(true);
            checkUrlInclude(herodotusTerminalPage.pageUrl);
            break;
        case EPagesName.tradingStats:
            tradingStatsPage.checkDailyVisibleTable();
            checkUrlInclude(tradingStatsPage.pageUrl);
            break;
        case EPagesName.backtestingManager:
            backtestingPage.checkVisiblePanel();
            checkUrlInclude(backtestingPage.pageUrl);
            break;
        case EPagesName.balanceMonitor:
            balanceMonitorPage.checkVisiblePanel();
            checkUrlInclude(balanceMonitorPage.pageUrl);
            break;
        case EPagesName.dashboard:
            dashboardPage.checkElementsExists();
            checkUrlInclude(dashboardPage.pageUrl);
            break;
        case EPagesName.jsonCharts:
            charterJsonViewerPage.checkElementsExists();
            checkUrlInclude(charterJsonViewerPage.pageUrl);
            break;
        case EPagesName.herodotusTrades:
            herodotusTradesPage.checkElementsExists();
            herodotusTradesPage.checkVisibleTable();
            checkUrlInclude(herodotusTradesPage.pageUrl);
            break;
        case EPagesName.wsQueryTerminal:
            wsQueryTerminalPage.checkElementsExists();
            wsQueryTerminalPage.checkVisibleTab();
            checkUrlInclude(wsQueryTerminalPage.pageUrl);
            break;
    }
});

Given(
    `user goes to the {string} page with the backend server parameter`,
    (pageName: EPagesName) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.openPageWithBackendServer();
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.openPageWithBackendServer();
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.openPageWithBackendServer();
                break;
            case EPagesName.backtestingManager:
                backtestingPage.openPageWithBackendServer();
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.openPageWithBackendServer();
                break;
            case EPagesName.wsQueryTerminal:
                wsQueryTerminalPage.openPageWithBackendServer();
                break;
        }
    },
);

Given(
    `user goes to the {string} page by {string} server params`,
    (pageName: EPagesName, nameServer: string) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.backtestingManager:
                backtestingPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.openPageByBackendServer(nameServer);
                break;
        }
    },
);

Given(
    `user goes to the "Backtesting Manager" page with the backend server {string}`,
    (nameServer: string) => {
        backtestingPage.openPageWithBackendServerByParams(nameServer);
    },
);

Given(`user goes to the "Herodotus Terminal" page by index {string}`, (index: string) => {
    herodotusTerminalPage.openPageWithBackendServerByIndex(index);
});

Given(`user goes to the "Herodotus Terminal" page by name {string}`, (robotName: string) => {
    const index = `${ERobots[robotName]}`;
    herodotusTerminalPage.openPageWithBackendServerByIndex(index);
});

Given(`user click "Log Out" button on the {string} page`, () => {
    authenticationPage.clickLogOutButton();
});

Given(`user on the {string} page with the "ModalSettings" modal`, (pageName: EPagesName) => {
    settingsModal.checkElementsExists();
    customWait(1);
    switch (pageName) {
        case EPagesName.tradingServersManager:
            checkUrlInclude(tradingServersManagerPage.pageUrl);
            break;
        case EPagesName.herodotusTerminal:
            checkUrlInclude(herodotusTerminalPage.pageUrl);
            break;
        case EPagesName.tradingStats:
            checkUrlInclude(tradingStatsPage.pageUrl);
            break;
        case EPagesName.backtestingManager:
            checkUrlInclude(backtestingPage.pageUrl);
            break;
    }
});

Given(`user opens the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.dashboard:
            dashboardPage.visit();
            dashboardPage.checkElementsExists();
            break;
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.tradingStats:
            tradingStatsPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.backtestingManager:
            backtestingPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.balanceMonitor:
            balanceMonitorPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.wsQueryTerminal:
            wsQueryTerminalPage.visit();
            settingsModal.checkElementsExists();
            break;
    }
});

Given(`user on the "Herodotus Terminal" page of the {string} robot`, (robotName: ERobots) => {
    checkUrlInclude(herodotusTerminalPage.pageUrl);
    herodotusTerminalPage.checkElementsExists();
    checkUrlInclude(`${ERobots[robotName]}`);
});

Given(
    `user types the modal settings with the {string} backend server parameter`,
    (backendServerName: string) => {
        if (backendServerName === 'default') {
            backendServerName = Cypress.env('backendServerName');
        }
        settingsModal.serverSelect.checkVisible();
        customWait(0.5);
        settingsModal.serverSelect.typeAndClickByText(backendServerName);
    },
);

Given(`user gets URL the "Charts" page and opens the page`, () => {
    tableRow.visitPageByLinkInRow();
});
