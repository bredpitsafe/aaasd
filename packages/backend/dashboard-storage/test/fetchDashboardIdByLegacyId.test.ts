import { afterAll, describe, expect, it } from 'vitest';

import { Client } from './client/index.js';
import importDashboard from './fixtures/importDashboard.json';

describe('FetchDashboardIdByLegacyId', () => {
    const client = new Client({ authenticate: true });

    afterAll(() => {
        client.close();
    });

    it('gets dashboard id by legacy id', () => async () => {
        const importResponse = await client.request(importDashboard);
        expect(importResponse.payload.id).toBeInstanceOf(String);

        const response = await client.request({
            type: 'FetchDashboardIdByLegacyId',
            legacyId: importDashboard.legacyId,
        });
        expect(response.state).toBe('Done');
        expect(response.payload).toEqual({
            type: 'DashboardId',
            id: importResponse.payload.id,
        });
    });

    it('fails with non-existent legacyId', async () => {
        const response = await client.request({
            type: 'FetchDashboardIdByLegacyId',
            legacyId: Math.floor(Math.random() * 1000000),
        });
        expect(response.state).toBe('Done');
        expect(response.error).toMatchSnapshot();
    });
});
