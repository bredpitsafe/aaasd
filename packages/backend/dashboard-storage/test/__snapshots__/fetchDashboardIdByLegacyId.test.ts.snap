// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FetchDashboardIdByLegacyId > fails with non-existent legacyId 1`] = `
{
  "description": "Dashboard not found: Dashboard does not exist in storage. Ask initial owner to manually upload it.",
  "kind": "Authorization",
}
`;
